plugins/backstage-plugin-coder/src/utils/StateSnapshotManager.ts

```diff
@@ -0,0 +1,166 @@
+/**
+ * @file A helper class that simplifies the process of connecting mutable class
+ * values (such as the majority of values from API factories) with React's
+ * useSyncExternalStore hook.
+ *
+ * This should not be used directly from within React, but should instead be
+ * composed into other classes (such as API factories). Those classes can then
+ * be brought into React.
+ *
+ * As long as you can figure out how to turn the mutable values in some other
+ * class into an immutable snapshot, all you have to do is pass the new snapshot
+ * into this class. It will then take care of notifying subscriptions, while
+ * reconciling old/new snapshots to minimize needless re-renders.
+ */
+import type { ReadonlyJsonValue } from '../typesConstants';
+
+type SubscriptionCallback<TSnapshot extends ReadonlyJsonValue> = (
+  snapshot: TSnapshot,
+) => void;
+
+type DidSnapshotsChange<TSnapshot extends ReadonlyJsonValue> = (
+  oldSnapshot: TSnapshot,
+  newSnapshot: TSnapshot,
+) => boolean;
+
+type SnapshotManagerOptions<TSnapshot extends ReadonlyJsonValue> = Readonly<{
+  initialSnapshot: TSnapshot;
+
+  /**
+   * Lets you define a custom comparison strategy for detecting whether a
+   * snapshot has really changed in a way that should be reflected in the UI.
+   */
+  didSnapshotsChange?: DidSnapshotsChange<TSnapshot>;
+}>;
+
+interface SnapshotManagerApi<TSnapshot extends ReadonlyJsonValue> {
+  subscribe: (callback: SubscriptionCallback<TSnapshot>) => () => void;
+  unsubscribe: (callback: SubscriptionCallback<TSnapshot>) => void;
+  getSnapshot: () => TSnapshot;
+  updateSnapshot: (newSnapshot: TSnapshot) => void;
+}
+
+function areSameByReference(v1: unknown, v2: unknown) {
+  // Comparison looks wonky, but Object.is handles more edge cases than ===
+  // for these kinds of comparisons, but it itself has an edge case
+  // with -0 and +0. Still need === to handle that comparison
```

Comment:
JavaScript is wild lol

Reply:
Yeah, I usually call `Object.is` the "quadruple equals" operator, but the flipside of that is that sometimes it gets too specific for its own good lol